Update newrelicexporter to use pdata for traces

MrAlias opened the pull request:
**Description:**  Update the newrelicexporter to use the `pdata` module instead of the OpenCensus proto for its tracing export pipeline. This includes updates to resource, instrumentation library, and status transformations to account for their form in the OTLP.

**Testing:** Unit tests have been updated to test the pdata data structures where possible.

MrAlias committed:
Update newrelicexporter to use pdata for traces

MrAlias committed:
Include instrumentation info

MrAlias committed:
Update transformer

Rely on the service name being set in the resource or the upstream span
instead of trying to parse something and add it to the ServiceName
field.

Fix tests

MrAlias committed:
Update transformer test

MrAlias committed:
Add tests for Span transform errors

MrAlias committed:
Abstract common attribute generation

MrAlias committed:
Add test for inst transform

MrAlias committed:
Lint var names

MrAlias commented:
@pyohannes @justinfoote I'm not able to request a review from you through the interface, but would appreciate your review.

MrAlias commented:
Not sure if the unit tests are transient or not? I didn't change anything in the k8sprocessor.

justinfoote reviewed (approved):
:rocket: Looks good to me!

MrAlias commented:
@anuraaga is there any chance you can re-run the tests to see if they were transient?

anuraaga commented on exporter/newrelicexporter/newrelic.go:
```diff
@@ -79,34 +79,44 @@ func newExporter(l *zap.Logger, c configmodels.Exporter) (*exporter, error) {
-	var errs []error
-	goodSpans := 0
-
-	octds := internaldata.TraceDataToOC(td)
```
🎉

anuraaga commented on exporter/newrelicexporter/newrelic.go:
```diff
@@ -79,34 +79,44 @@ func newExporter(l *zap.Logger, c configmodels.Exporter) (*exporter, error) {
+
+				nrSpan, err := transform.Span(span)
+				if err != nil {
+					// Record the error and try export anyway to help debug.
```
What gets exported if the conversion failed? Partially transformed span? Seems a bit scary

MrAlias replied:
Yeah, we can probably just continue like before. The two situations where it would return a partial span (outside of a nil span) are a bad trace ID or bad span ID which the harvester will return an error:

https://github.com/newrelic/newrelic-telemetry-sdk-go/blob/e3d912863ace96681b2b66ebffdb0acc5f874ff2/telemetry/harvester.go#L106-L116

This means that nothing will be exported, but error will be duplicated.

I'll update.

anuraaga commented on exporter/newrelicexporter/transformer.go:
```diff
@@ -22,124 +22,134 @@ import (
-	if isErr {
-		attrs["error"] = true
+	if hasStatusCode {
+		code := strings.TrimPrefix(span.Status().Code().String(), "STATUS_CODE_")
```
Consider a switch statement for some performance improvement here and for Kind

MrAlias replied:
I'm not sure I follow. Each one of these conditional statements is independent. They each have unique clauses that are decoupled from one another.

To replicate the same behavior they would be transformed to a switch statement with one case. Is that what you envisioned?

anuraaga reviewed (approved):
Just small comments, thanks

anuraaga commented:
Also would be great if you can check out the codecov report, looks like we're missing some tests.

MrAlias committed:
Do not try to export partial spans

MrAlias commented:
> Also would be great if you can check out the codecov report, looks like we're missing some tests

Yeah, I looked into adding tests for when the pdata structs are nil, but given the OTLP is not exported from the collector and the internal data struct of not directly available in the pdata lib I'm not able to make certain Spans, InstrumentationLibrarySpans, or ResourceSpans nil. I'll take another look though.

MrAlias committed:
Update tests

Add test for no trace/span id from exporter.
Add test for merge attrs error.
Add factory tests.
